Patched JAR does not start when a commons JAR is updated

Spring Boot requires JARs inside the fat JAR to be stored without compression.

`Caused by: java.lang.IllegalStateException: Unable to open nested entry 'BOOT-INF/lib/zowe-rest-api-commons-spring-0.0.0-SNAPSHOT.jar'. It has been compressed and nested jar files must be stored without compression. Please check the mechanism used to create your executable jar file`

The workaround is to use:

```
zowe-api-dev deploy --force
```

Full error log:

```
Starting application in SSH z/OS UNIX session using command '/sys/java64bt/v8r0m0/usr/lpp/java/J8.0_64/bin/java -Djava.library.path="./lib:${LIBPATH}" -Xquickstart -jar bin/zowe-rest-api-sample-spring.jar --spring.config.additional-location=file:etc/application.yml' in directory '/a/plape03/zowe-rest-api-sample-spring'
ℹ You can stop it using Ctrl+C
Executing z/OS UNIX command '/sys/java64bt/v8r0m0/usr/lpp/java/J8.0_64/bin/java -Djava.library.path="./lib:${LIBPATH}" -Xquickstart -jar bin/zowe-rest-api-sample-spring.jar --spring.config.additional-location=file:etc/application.yml' in directory /a/plape03/zowe-rest-api-sample-spring

$ Exception in thread "main" java.lang.IllegalStateException: Failed to get nested archive for entry BOOT-INF/lib/zowe-rest-api-commons-spring-0.0.0-SNAPSHOT.jar
        at org.springframework.boot.loader.archive.JarFileArchive.getNestedArchive(JarFileArchive.java:108)
        at org.springframework.boot.loader.archive.JarFileArchive.getNestedArchives(JarFileArchive.java:87)
        at org.springframework.boot.loader.ExecutableArchiveLauncher.getClassPathArchives(ExecutableArchiveLauncher.java:69)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
        at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:52)
Caused by: java.io.IOException: Unable to open nested jar file 'BOOT-INF/lib/zowe-rest-api-commons-spring-0.0.0-SNAPSHOT.jar'
        at org.springframework.boot.loader.jar.JarFile.getNestedJarFile(JarFile.java:258)
        at org.springframework.boot.loader.jar.JarFile.getNestedJarFile(JarFile.java:244)
        at org.springframework.boot.loader.archive.JarFileArchive.getNestedArchive(JarFileArchive.java:104)
        ... 4 more
Caused by: java.lang.IllegalStateException: Unable to open nested entry 'BOOT-INF/lib/zowe-rest-api-commons-spring-0.0.0-SNAPSHOT.jar'. It has been compressed and nested jar files must be stored without compression. Please check the mechanism used to create your executable jar file
        at org.springframework.boot.loader.jar.JarFile.createJarFileFromFileEntry(JarFile.java:284)
        at org.springframework.boot.loader.jar.JarFile.createJarFileFromEntry(JarFile.java:266)
        at org.springframework.boot.loader.jar.JarFile.getNestedJarFile(JarFile.java:255)
        ... 6 more
Error: empty JSON response from Zowe CLI
    at zoweSync (~/workspace/github.com/zowe/sample-spring-boot-api-service/zowe-api-dev/src/zowe.ts:65:23)
    at execSshCommand (~/workspace/github.com/zowe/sample-spring-boot-api-service/zowe-api-dev/src/zowe.ts:152:12)
    at Object.execSshCommandWithDefaultEnv (~/workspace/github.com/zowe/sample-spring-boot-api-service/zowe-api-dev/src/zowe.ts:161:12)
    at Start.run (~/workspace/github.com/zowe/sample-spring-boot-api-service/zowe-api-dev/src/commands/start.ts:59:13)
    at Start._run (~/workspace/github.com/zowe/sample-spring-boot-api-service/zowe-api-dev/node_modules/@oclif/command/lib/command.js:44:31)
```

--- a/jarpatcher/src/main/java/jarpatcher/JarPatcher.java
+++ b/jarpatcher/src/main/java/jarpatcher/JarPatcher.java
@@ -16,17 +16,14 @@ import java.io.File;
 import java.io.FileInputStream;
 import java.io.FileOutputStream;
 import java.io.IOException;
-import java.net.URI;
-import java.nio.file.FileSystem;
-import java.nio.file.FileSystems;
+import java.io.InputStream;
+import java.io.OutputStream;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.nio.file.StandardCopyOption;
 import java.util.Enumeration;
-import java.util.HashMap;
 import java.util.HashSet;
-import java.util.Map;
 import java.util.Set;
 import java.util.logging.Logger;
 import java.util.stream.Collectors;
@@ -72,7 +69,7 @@ public class JarPatcher {
             } else if (args[0].equals("patch")) {
                 String targetPath = args[1];
                 String patchPath = args[2];
-                String ignoredPrefix = (args.length >= 4) ?  args[3] : null;
+                String ignoredPrefix = (args.length >= 4) ? args[3] : null;
                 applyPatch(targetPath, patchPath, ignoredPrefix);
             }
         } catch (IOException e) {
@@ -109,13 +106,15 @@ public class JarPatcher {
     }
 
     private void printEntry(ZipFile f, ZipEntry entry) {
-        logger.info(String.format("zipFile=%s file=%s size=%d crc=%d dir=%b comment=%s", f.getName(), entry.getName(),
-                entry.getSize(), entry.getCrc(), entry.isDirectory(), entry.getComment()));
+        logger.info(String.format("zipFile=%s file=%s size=%d crc=%d dir=%b comment=%s method=%s", f.getName(),
+                entry.getName(), entry.getSize(), entry.getCrc(), entry.isDirectory(), entry.getComment(),
+                entry.getMethod()));
     }
 
     private boolean compareEntries(ZipEntry entry1, ZipEntry entry2) {
         return (entry1.isDirectory() == entry2.isDirectory()) && (entry1.getSize() == entry2.getSize())
-                && (entry1.getName().equals(entry2.getName())) && (entry1.getCrc() == entry2.getCrc());
+                && (entry1.getName().equals(entry2.getName())) && (entry1.getCrc() == entry2.getCrc()
+                        && (entry1.isDirectory() || (entry1.getMethod() == entry2.getMethod())));
     }
 
     private Set<String> setFromEntryNames(ZipFile zipFile) {
@@ -136,19 +135,24 @@ public class JarPatcher {
             String filename = inEntry.getName();
             if (result.isFileChangedOrCreated(filename)) {
                 ZipEntry patchEntry = new ZipEntry(filename);
+                copyZipEntryAttributes(inEntry, patchEntry);
                 zipOut.putNextEntry(patchEntry);
                 if (!inEntry.isDirectory()) {
-                    byte[] buffer = new byte[1024];
-                    int len;
-                    while ((len = zipIn.read(buffer)) > 0) {
-                        zipOut.write(buffer, 0, len);
-                    }
+                    copyStream(zipIn, zipOut);
                 }
                 zipOut.closeEntry();
             }
             inEntry = zipIn.getNextEntry();
         }
         zipIn.close();
+    }
+
+    private void copyStream(InputStream zipIn, OutputStream zipOut) throws IOException {
+        byte[] buffer = new byte[4096];
+        int len;
+        while ((len = zipIn.read(buffer)) > 0) {
+            zipOut.write(buffer, 0, len);
+        }
     }
 
     private void addDeletedEntries(Set<String> deletedFiles, ZipOutputStream zipOut) throws IOException {
@@ -159,8 +163,7 @@ public class JarPatcher {
         }
     }
 
-    private void addFiles(String fromPath, ZipOutputStream zipOut, String ignoredPath)
-            throws IOException {
+    private void addFiles(String fromPath, ZipOutputStream zipOut, String ignoredPath) throws IOException {
         ZipInputStream zipIn = new ZipInputStream(new FileInputStream(fromPath));
         ZipEntry inEntry = zipIn.getNextEntry();
         while (inEntry != null) {
@@ -182,7 +185,8 @@ public class JarPatcher {
         zipIn.close();
     }
 
-    public boolean createPatch(String oldPath, String newPath, String patchPath, String patcherPath) throws ZipException, IOException {
+    public boolean createPatch(String oldPath, String newPath, String patchPath, String patcherPath)
+            throws ZipException, IOException {
         CompareResult result = compare(oldPath, newPath);
         logger.info(String.format("Comparing %s and %s: %s", oldPath, newPath, result));
 
@@ -201,30 +205,26 @@ public class JarPatcher {
         return result.archivesAreSame();
     }
 
-    private static String operatingSystem = System.getProperty("os.name");
-
-    private static boolean isWindows() {
-       return operatingSystem.startsWith("Windows");
-    }
-
-    private URI jarFileUri(String path) {
-        if (isWindows()) {
-            return URI.create("jar:file:/" + path.replace("\\", "/"));
+    private void createDirectories(String filename, ZipOutputStream zipOut, Set<String> createdDirectories)
+            throws IOException {
+        Path dir = Paths.get(filename).getParent();
+        String dirname = (dir == null) ? null : (dir.toString() + "/");
+        if ((dirname != null) && !createdDirectories.contains(dirname)) {
+            createDirectories(dirname, zipOut, createdDirectories);
+            ZipEntry zipEntry = new ZipEntry(dirname);
+            zipOut.putNextEntry(zipEntry);
+            zipOut.closeEntry();
+            createdDirectories.add(dirname);
         }
-        return URI.create("jar:file://" + path);
     }
 
     public void applyPatch(String targetPath, String patchPath, String ignoredPath) throws IOException {
-        FileSystem jarPatch;
-        FileSystem jarTarget;
-        Map<String, String> env = new HashMap<>();
-        env.put("create", "false");
-        try {
-            jarPatch = FileSystems.newFileSystem(jarFileUri(patchPath), env);
-            jarTarget = FileSystems.newFileSystem(jarFileUri(targetPath), env);
-        } catch (IllegalArgumentException e) {
-            throw new RuntimeException(e);
-        }
+        String originalTargetPath = targetPath + "-original";
+        Files.copy(Paths.get(targetPath), Paths.get(originalTargetPath), StandardCopyOption.REPLACE_EXISTING);
+        FileOutputStream fos = new FileOutputStream(targetPath.toString());
+        ZipOutputStream zipOut = new ZipOutputStream(fos);
+        Set<String> deletedOrPatchedNames = new HashSet<>();
+
         ZipFile zipPatch = new ZipFile(patchPath);
         Enumeration<? extends ZipEntry> entries = zipPatch.entries();
         while (entries.hasMoreElements()) {
@@ -234,23 +234,54 @@ public class JarPatcher {
             if (filename.startsWith(DELETED)) {
                 String realFilename = filename.substring(DELETED.length());
                 logger.info("Deleting: " + realFilename);
-                Files.delete(jarTarget.getPath(realFilename));
+                deletedOrPatchedNames.add(realFilename);
             } else if ((ignoredPath != null) && (filename.startsWith(ignoredPath))) {
                 logger.info("Ignoring: " + filename);
             } else {
                 logger.info("Patching: " + filename);
-                Path dir = Paths.get(filename).getParent();
-                String dirname = (dir == null) ? null : dir.toString();
-                if ((dirname != null) && !Files.exists(jarTarget.getPath(dirname))) {
-                    Files.createDirectories(jarTarget.getPath(dirname));
-                }
-                Files.copy(jarPatch.getPath(filename), jarTarget.getPath(filename),
-                        StandardCopyOption.REPLACE_EXISTING);
+                deletedOrPatchedNames.add(filename);
+                writeEntry(zipOut, deletedOrPatchedNames, zipPatch, entry, filename);
             }
         }
-        jarTarget.close();
         zipPatch.close();
-        jarPatch.close();
+
+        ZipFile zipIn = new ZipFile(originalTargetPath);
+        entries = zipIn.entries();
+        while (entries.hasMoreElements()) {
+            ZipEntry entry = entries.nextElement();
+            String filename = entry.getName();
+
+            if (!deletedOrPatchedNames.contains(filename)) {
+                logger.info("Keeping: " + filename);
+                deletedOrPatchedNames.add(filename);
+                writeEntry(zipOut, deletedOrPatchedNames, zipIn, entry, filename);
+            }
+        }
+        zipIn.close();
+
+        zipOut.close();
+        fos.close();
+        Files.delete(Paths.get(originalTargetPath));
+    }
+
+    private void writeEntry(ZipOutputStream zipOut, Set<String> createdDirectories, ZipFile zipPatch, ZipEntry entry,
+            String filename) throws IOException {
+        createDirectories(filename, zipOut, createdDirectories);
+        ZipEntry zipEntry = new ZipEntry(filename);
+        copyZipEntryAttributes(entry, zipEntry);
+        InputStream inputStream = zipPatch.getInputStream(entry);
+        zipOut.putNextEntry(zipEntry);
+        copyStream(inputStream, zipOut);
+        zipOut.closeEntry();
+    }
+
+    private void copyZipEntryAttributes(ZipEntry inEntry, ZipEntry outEntry) {
+        outEntry.setMethod(inEntry.getMethod());
+        outEntry.setComment(inEntry.getComment());
+        if (inEntry.getMethod() == ZipEntry.STORED) {
+            outEntry.setCrc(inEntry.getCrc());
+            outEntry.setCompressedSize(inEntry.getCompressedSize());
+        }
     }
 
     public static void main(String[] args) {

--- a/jarpatcher/src/test/java/jarpatcher/JarPatcherTests.java
+++ b/jarpatcher/src/test/java/jarpatcher/JarPatcherTests.java
@@ -23,6 +23,7 @@ import java.nio.file.Paths;
 import java.util.HashSet;
 import java.util.Set;
 import java.util.logging.Logger;
+import java.util.zip.CRC32;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipException;
 import java.util.zip.ZipOutputStream;
@@ -84,6 +85,13 @@ public class JarPatcherTests {
     }
 
     @Test
+    public void jarFilesCanBeAddedToDirectoryInArchive() throws ZipException, IOException {
+        testPatchFromArchiveToAnotherArchive(new ArchiveDefinition().addFiles(new TestFile("dir/one.jar", SOME_DATA)),
+                new ArchiveDefinition().addFiles(new TestFile("dir/one.jar", SOME_DATA),
+                        new TestFile("dir/second.jar", SOME_DATA)));
+    }
+
+    @Test
     public void directoriesCanBeAddedToEmptyArchive() throws ZipException, IOException {
         testPatchFromArchiveToAnotherArchive(new ArchiveDefinition(),
                 new ArchiveDefinition().addFiles(new TestFile("dir/")));
@@ -124,6 +132,13 @@ public class JarPatcherTests {
             String filename = f.getFilename();
             createDirectories(filename, zipOut, createdDirectories);
             ZipEntry zipEntry = new ZipEntry(filename);
+            zipEntry.setMethod(filename.endsWith(".jar") ? ZipEntry.STORED : ZipEntry.DEFLATED);
+            if ((zipEntry.getMethod() == ZipEntry.STORED) && (f.getData() != null)) {
+                zipEntry.setCompressedSize(f.getData().getBytes().length);
+                CRC32 crc = new CRC32();
+                crc.update(f.getData().getBytes());
+                zipEntry.setCrc(crc.getValue());
+            }
             zipOut.putNextEntry(zipEntry);
             if (f.getData() != null) {
                 zipOut.write(f.getData().getBytes());
@@ -159,9 +174,9 @@ public class JarPatcherTests {
         logger.info("Old JAR: " + oldDef.toString());
         logger.info("New JAR: " + newDef.toString());
         JarPatcher jarPatcher = new JarPatcher();
-        String oldPath = createTestArchiveFromDefinition("a", oldDef);
-        String oldPath2 = createTestArchiveFromDefinition("a2", oldDef);
-        String newPath = createTestArchiveFromDefinition("b", newDef);
+        String oldPath = createTestArchiveFromDefinition("old", oldDef);
+        String oldPath2 = createTestArchiveFromDefinition("old2", oldDef);
+        String newPath = createTestArchiveFromDefinition("new", newDef);
         String patchPath = tempDir.resolve("patch.jar").toString();
 
         String patcherPath = null;
@@ -183,11 +198,13 @@ public class JarPatcherTests {
         jarPatcher.applyPatch(oldPath, patchPath, ignoredPath);
         CompareResult compareResult = jarPatcher.compare(oldPath, newPath);
         logger.info("Compare result: " + compareResult.toString());
-        assertTrue(compareResult.archivesAreSame());
+        assertTrue("patched file is not same as the target: " + compareResult.toString(),
+                compareResult.archivesAreSame());
 
         new JarPatcher().run(new String[] { "patch", oldPath2, patchPath, ignoredPath });
         CompareResult compareResult2 = jarPatcher.compare(oldPath2, newPath);
         logger.info("Compare result: " + compareResult2.toString());
-        assertTrue(compareResult2.archivesAreSame());
+        assertTrue("patched file is not same as the target (via JarPatcher): " + compareResult2.toString(),
+                compareResult2.archivesAreSame());
     }
 }

--- a/scripts/run-integration-tests-zos.sh
+++ b/scripts/run-integration-tests-zos.sh
@@ -2,7 +2,7 @@
 scripts/lock-port.sh
 
 cd zowe-rest-api-sample-spring
-zowe-api-dev deploy --force
+zowe-api-dev deploy
 zowe-api-dev config --name zos -p port=$TEST_PORT
 zowe-api-dev start --job
 

--- a/zowe-api-dev/src/commands/deploy.ts
+++ b/zowe-api-dev/src/commands/deploy.ts
@@ -8,12 +8,13 @@ export default class Deploy extends Command {
     static description = "deploy the API service artifacts to z/OS";
 
     static flags = {
-        force: flags.boolean({ char: "f", description: "forces full deployment even if there is not change", default: true }),
+        force: flags.boolean({ char: "f", description: "forces full deployment even if there is not change", default: false }),
     };
 
     async run() {
         const f = this.parse(Deploy).flags;
         const [userConfig, projectConfig] = readConfiguration(this);
+
 
         checkZowe(this);
 

--- a/zowe-api-dev/src/files.ts
+++ b/zowe-api-dev/src/files.ts
@@ -7,7 +7,7 @@ import { dirname, resolve } from "path";
 import * as tmp from "tmp";
 import { Entry, Parse } from "unzipper";
 import { checkZoweProfileName, ITransferredFile, IUserConfig } from "./config";
-import { execSshCommandWithDefaultEnv, execSshCommandWithDefaultEnvCwd, zoweSync } from "./zowe";
+import { execSshCommandWithDefaultEnv, execSshCommandWithDefaultEnvCwd, IApiResponse, zoweSync } from "./zowe";
 
 const debug = Debug("files");
 const userDataDir = ".zowe-api-dev";
@@ -46,6 +46,16 @@ export function isFileSame(file: string, zosFile: string, profileName: string): 
     return false;
 }
 
+export function zosExistsSync(zosFile: string): boolean {
+    const data = zoweSync(`files list uss ${zosFile}`, { throwError: false, logOutput: false }).data as IApiResponse;
+    try {
+        return data.apiResponse.totalRows === 1;
+    }
+    catch (e) {
+        return false;
+    }
+}
+
 export function transferFiles(
     files: { [filename: string]: ITransferredFile },
     zosTargetDir: string,
@@ -74,12 +84,13 @@ export function transferFiles(
             command.log(`Uploading template ${file} to ${zosFile}`);
             zoweSync(`files upload ftu ${tmpPath} ${zosFile}${options.binary ? " --binary" : ""}`);
         } else {
-            if (!force && isFileSame(file, zosFile, userConfig.zoweProfileName)) {
+            const zosFileExists = !force && zosExistsSync(zosFile);
+            if (!force && zosFileExists && isFileSame(file, zosFile, userConfig.zoweProfileName)) {
                 command.log(`${file} has not changed`)
                 return;
             }
             const oldFile = cachedOldFilePath(zosFile, userConfig.zoweProfileName);
-            if (!force && file.endsWith(".jar") && existsSync(oldFile)) {
+            if (!force && file.endsWith(".jar") && existsSync(oldFile) && zosFileExists) {
                 command.log(`Patching ${zosFile} to be same as ${file}`);
                 const patchFile = file + "-patch";
                 const zosPatchFile = zosFile + "-patch";

--- a/zowe-api-dev/src/zowe.ts
+++ b/zowe-api-dev/src/zowe.ts
@@ -19,6 +19,15 @@ export interface IZoweOptions {
     throwError?: boolean;
 }
 
+export interface IApiResponse {
+    apiResponse: {
+        items: [];
+        totalRows: number;
+        returnedRows: number;
+        JSONversion: number;
+    }
+}
+
 export function zoweSync(command: string, options?: IZoweOptions): IZoweResult {
     const defaultOptions: IZoweOptions = { direct: false, logOutput: true, throwError: true };
     if (options === undefined) {

